### ISSUE TITLE
-DIREE_BUILD_PYTHON_BINDINGS=ON build fail

### ISSUE THREAD

**Issue by hunbssfy** [google/iree#6801]
**Describe the bug**
Following the official https://google.github.io/iree/building-from-source/getting-started/ guide here, If enable python bindings, the build process fails.

**To Reproduce**
Follow the official build guide here: https://google.github.io/iree/building-from-source/getting-started/
When configuring cmake, add -DIREE_BUILD_PYTHON_BINDINGS=ON option like this:
`cmake -B ../iree-build/ -DCMAKE_BUILD_TYPE=RelWithDebInfo -DIREE_BUILD_PYTHON_BINDINGS=ON .
`
The build will fail at around 98%:
![image](https://user-images.githubusercontent.com/6930027/129999028-b8282529-e573-4dcc-a3de-70d58b2b99a7.png)

If I am building the python bindings the wrong way, please point out, thanks!

**Comment by GMNGeoffrey** [comment]
These are the instructions for building Python: https://google.github.io/iree/building-from-source/optional-features/#building-python-bindings

That said, the only difference I see between what you're doing and those instructions is it suggests explicitly specifying the Python3 executable and I don't see how failing to do so could cause that error.

I was just able to reproduce:

```shell
$ rm -rf ~/build/iree/bindings/
$ cmake -B ~/build/iree/bindings/ -DCMAKE_BUILD_TYPE=RelWithDebInfo -DIREE_BUILD_PYTHON_BINDINGS=ON
$ cmake --build ~/build/iree/bindings/ -j"$(nproc)"
...
[ 98%] Built target iree_vm_test_emitc_arithmetic_ops_gen
CMake Error: failed to create symbolic link '/usr/local/google/home/gcmn/build/iree/bindings/bindings/python/iree/tools/core/scripts/iree_translate/__main__.py': no such file or directory
[ 98%] Generating add_static.vmfb
gmake[2]: *** [bindings/python/iree/tools/core/CMakeFiles/bindings_python_iree_tools_core_core.dir/build.make:71: bindings_python_iree_tools_core_core] Error 1
[ 98%] Generating add_dynamic.vmfb
[ 98%] Built target iree_test_e2e_xla_ops_check_vulkan-spirv_vulkan_log.mlir_module
gmake[1]: *** [CMakeFiles/Makefile2:89958: bindings/python/iree/tools/core/CMakeFiles/bindings_python_iree_tools_core_core.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
...
```

But I don't get the same error with Ninja:

```shell
$ mv ~/build/iree/bindings ~/build/iree/bindings-make
$ cmake -GNinja -B ~/build/iree/bindings/ -DCMAKE_BUILD_TYPE=RelWithDebInfo -DIREE_BUILD_PYTHON_BINDINGS=ON
$ cmake --build ~/build/iree/bindings/
```

Basically everyone on the team and the CI use `ninja`, so we've had a few issues before where there are failures only for `make`. I don't understand *why* that is, but as a workaround for now, could you try using Ninja (install https://ninja-build.org/ and then `-GNinja`)? We're almost at the point of just listing that as a requirement (or as the only supported generator). Just generally you're likely to have a nicer time with that anyway :-)

**Comment by GMNGeoffrey** [comment]
That error suggests to me that the *directory* it's trying to create a symlink in doesn't exist. You can happily create a symlink to something that doesn't exist (try `ln -s doesnotexist foobar`, but just like you can't create a file in a directory that doesn't already exist, you can't create symlink there and it will give exactly that error (which leaves something to be desired in terms of quality error messages):

```shell
$ ln -s doesexist does/not/exist
ln: failed to create symbolic link 'does/not/exist': No such file or directory
```

And indeed, that directory isn't there in my make build directory:

```shell
$ ls ~/build/iree/bindings-make/bindings/python/iree/tools/core/
CMakeFiles  cmake_install.cmake  CTestTestfile.cmake  __init__.py  Makefile
```

but is in my ninja build directory:


```shell
$ ls ~/build/iree/bindings-ninja/bindings/python/iree/tools/core/
CMakeFiles  cmake_install.cmake  CTestTestfile.cmake  __init__.py  iree-translate  scripts
```

It's *not* right after I configure, so something in the build is creating this directory, and I guess in ninja it either correctly determines the dependency or happens to occur first:

```shell
$ cmake -GNinja -B ~/build/iree/bindings/ -DCMAKE_BUILD_TYPE=RelWithDebInfo -DIREE_BUILD_PYTHON_BINDINGS=ON
$ ls ~/build/iree/bindings/bindings/python/iree/tools/core/
CMakeFiles  cmake_install.cmake  CTestTestfile.cmake
```


```shell
$ grep -C 7 -n /scripts ~/build/iree/bindings-ninja/build.ninja
...
151956-#############################################
151957-# Utility command for bindings_python_iree_tools_core_core
151958-
151959:build bindings/python/iree/tools/core/CMakeFiles/bindings_python_iree_tools_core_core.util bindings/python/iree/tools/core/__init__.py bindings/python/iree/tools/core/scripts/iree_translate/__main__.py bindings/python/iree/tools/core/iree-translate: CUSTOM_COMMAND iree/tools/iree-translate
151960:  COMMAND = cd /usr/local/google/home/gcmn/build/iree/bindings/bindings/python/iree/tools/core && /usr/local/google/home/gcmn/bin/cmake -E create_symlink /usr/local/google/home/gcmn/src/iree/bindings/python/iree/tools/core/__init__.py /usr/local/google/home/gcmn/build/iree/bindings/bindings/python/iree/tools/core/__init__.py && cd /usr/local/google/home/gcmn/build/iree/bindings/bindings/python/iree/tools/core && /usr/local/google/home/gcmn/bin/cmake -E create_symlink /usr/local/google/home/gcmn/src/iree/bindings/python/iree/tools/core/scripts/iree_translate/__main__.py /usr/local/google/home/gcmn/build/iree/bindings/bindings/python/iree/tools/core/scripts/iree_translate/__main__.py && cd /usr/local/google/home/gcmn/build/iree/bindings/bindings/python/iree/tools/core && /usr/local/google/home/gcmn/bin/cmake -E create_symlink /usr/local/google/home/gcmn/build/iree/bindings/iree/tools/iree-translate /usr/local/google/home/gcmn/build/iree/bindings/bindings/python/iree/tools/core/iree-translate
151961-  DESC = Running utility command for bindings_python_iree_tools_core_core
151962-  restat = 1
151963-
151964-build bindings/python/iree/tools/core/bindings_python_iree_tools_core_core: phony bindings/python/iree/tools/core/CMakeFiles/bindings_python_iree_tools_core_core.util
151965-
151966-
151967-#############################################
```

vs

```
$ grep /scripts ~/build/iree/bindings-make/bindings/python/iree/tools/core/CMakeFiles/bindings_python_iree_tools_core_core.dir/build.make 
        cd /usr/local/google/home/gcmn/build/iree/bindings/bindings/python/iree/tools/core && /usr/local/google/home/gcmn/bin/cmake -E create_symlink /usr/local/google/home/gcmn/src/iree/bindings/python/iree/tools/core/scripts/iree_translate/__main__.py /usr/local/google/home/gcmn/build/iree/bindings/bindings/python/iree/tools/core/scripts/iree_translate/__main__.py
```

**Comment by GMNGeoffrey** [comment]
Let's figure out what creates the scripts directory during the ninja build:

```shell
$ sudo touch ~/build/iree/bindings/bindings/python/iree/tools/core/scripts
$ cmake --build ~/build/iree/bindings/
...
[3919/4467] Running utility command for bindings_python_iree_tools_core_coreninja: error: mkdir(bindings/python/iree/tools/core/scripts/iree_translate): Not a directory
```

So it seems that ninja is creating that directory, but honestly I'm not sure what's telling it to do so...

**Comment by GMNGeoffrey** [comment]
Much grepping later, I'm none the wiser. I don't know where this bit of ninja configuration happens

**Comment by stellaraccident** [comment]
Passed each other on vacations and i didn't have time to land did updates: https://github.com/google/iree/pull/6696

Should probably just explicitly create those directories at configure time in cmake.

**Comment by GMNGeoffrey** [comment]
Yeah it seems that somehow ninja figures out that it needs to mkdir before creating an output here and make doesn't. I poked through the rules and don't see any mention of an explicit mkdir. `cmake -E create_symlink doesexist does/not/exist` fails as expected, so it's not just that. My Google searches are not turning up any results

**Comment by GMNGeoffrey** [comment]
This bit of code in Ninja looks like it. Seems Ninja just always creates the directory for outputs:  https://github.com/ninja-build/ninja/blob/ffb47fcf82ed12f53c046f6d51401c34ec35eced/src/build.cc#L682-L688

Anyway, #6808 should fix this by adding explicit directory creation to the places where we create symlinks that might be in directories that don't exist

**Comment by hunbssfy** [comment]
Hi Geroffery,
Ninja fix the problem, thanks!
